File: Blazor/Core/Services/RiskMatrixService.cs
namespace Construction.Blazor.Core.Services;

/// <summary>Flattened view model for the Probability x Severity risk matrix — the server
/// returns a nested rows/cells shape (see RiskMatrixDto) with probability/severity as plain
/// strings; this is easier to look up by (probability, severity) pair when rendering the grid.</summary>
public record RiskMatrixCellViewModel(string Probability, string Severity, int Count, IReadOnlyList<string> RiskIds);

public class RiskMatrixService(RisksService risks)
{
    public async Task<List<RiskMatrixCellViewModel>> GetMatrixAsync()
    {
        var matrix = await risks.GetMatrixAsync();
        return matrix.Rows
            .SelectMany(row => row.Cells.Select(cell =>
                new RiskMatrixCellViewModel(row.Probability, cell.Severity, cell.RiskNumbers.Count, cell.RiskNumbers)))
            .ToList();
    }
}

File: Angular/src/app/core/services/risk-matrix.service.ts
import { Injectable, inject } from '@angular/core';
import { Observable, map } from 'rxjs';
import { ApiClientService } from './api-client.service';
import type { RiskMatrixCellViewModel, RiskMatrixDto } from '../models/api.models';

function flattenMatrix(matrix: RiskMatrixDto): RiskMatrixCellViewModel[] {
  return matrix.rows.flatMap((row) =>
    row.cells.map((cell) => ({
      probability: row.probability,
      severity: cell.severity,
      count: cell.riskNumbers.length,
      riskIds: cell.riskNumbers,
    })),
  );
}

@Injectable({ providedIn: 'root' })
export class RiskMatrixService {
  private api = inject(ApiClientService);

  getMatrix(): Observable<RiskMatrixCellViewModel[]> {
    return this.api.getJson<RiskMatrixDto>('risks/matrix').pipe(map(flattenMatrix));
  }
}
